internal/controllers/phase_ensure_operator_group.go

```diff
@@ -0,0 +1,104 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	operatorsv1 "github.com/operator-framework/api/pkg/operators/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	addonsv1alpha1 "github.com/openshift/addon-operator/apis/addons/v1alpha1"
+)
+
+// Ensures the presense or absense of an OperatorGroup depending on the Addon install type.
+func (r *AddonReconciler) ensureOperatorGroup(
+	ctx context.Context, log logr.Logger, addon *addonsv1alpha1.Addon) (stop bool, err error) {
+	switch addon.Spec.Install.Type {
+	case addonsv1alpha1.AllNamespaces:
+		// No OperatorGroup needed, Operator is installed for the whole cluster
+		// Ensure that we have no left over OperatorGroup.
+		operatorGroupList := &operatorsv1.OperatorGroupList{}
+		err := r.List(ctx, operatorGroupList, client.MatchingLabelsSelector{
+			Selector: commonLabelsAsLabelSelector(addon),
+		})
+		if err != nil {
+			return false, err
+		}
+
+		for _, operatorGroup := range operatorGroupList.Items {
+			if err := client.IgnoreNotFound(r.Delete(ctx, &operatorGroup)); err != nil {
+				return false, err
+			}
+		}
+
+		return false, nil
+
+	case addonsv1alpha1.OwnNamespaces:
+		// continue
+
+	default:
+		// Unsupported Install Type
+		// This should never happen, unless the schema validation is wrong.
+		// The .install.type property is set to only allow known enum values.
+		log.Error(fmt.Errorf("invalid Addon install type: %q", addon.Spec.Install.Type), "stopping Addon reconcilation")
+		return true, nil
+	}
+
+	if addon.Spec.Install.OwnNamespace == nil ||
+		len(addon.Spec.Install.OwnNamespace.Namespace) == 0 {
+		// invalid/missing configuration
+		// TODO: Move error reporting into webhook and reduce this code to a sanity check.
+		addon.Status.ObservedGeneration = addon.Generation
+		addon.Status.Phase = addonsv1alpha1.PhaseError
+		meta.SetStatusCondition(&addon.Status.Conditions, metav1.Condition{
+			Type:    addonsv1alpha1.Available,
+			Status:  metav1.ConditionFalse,
+			Reason:  "ConfigurationError",
+			Message: ".spec.install.ownNamespace.namespace is required when .spec.install.type = OwnNamespace",
+		})
+		return true, r.Status().Update(ctx, addon)
+	}
+
+	desiredOperatorGroup := &operatorsv1.OperatorGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      addon.Name,
+			Namespace: addon.Spec.Install.OwnNamespace.Namespace,
+			Labels:    map[string]string{},
+		},
+		Spec: operatorsv1.OperatorGroupSpec{
+			TargetNamespaces: []string{addon.Spec.Install.OwnNamespace.Namespace},
```

Comment:
Thinking abour the API here, both AllNamespaces and OwnNamespace will make the OperatorGroup be installed in what we consider the "addon target namespace". "targetNamespace" should probably be a top level parameter in the `install` section, so we never have to list namespace based on the install type. does that make sense?

Reply:
```
apiVersion: addons.managed.openshift.io/v1alpha1
kind: Addon
metadata:
  name: example-addon
spec:
  namespaces:
  - name: example-namespace
  install:
    targetNamespace: example-namespace
    type: AllNamespaces
```

Resulting in 

```
apiVersion: operators.coreos.com/v1alpha2
kind: OperatorGroup
metadata:
  name: example-operator-group
  namespace: example-namespace
```

and 

```
apiVersion: addons.managed.openshift.io/v1alpha1
kind: Addon
metadata:
  name: example-addon
spec:
  namespaces:
  - name: example-namespace
  install:
    targetNamespace: example-namespace
    type: OwnNamespace
```
resulting in:

```
apiVersion: operators.coreos.com/v1alpha2
kind: OperatorGroup
metadata:
  name: redhat-layered-product-og
  namespace: example-namespace
spec:
  targetNamespaces:
  - example-namespace
```

Reply:
+1 for option 1... I love being future proof :)
Keep in mind that the namespace there will also be used for creating the Subscription.

Reply:
Implemented